### DISCUSSION TITLE
Why do we fuzz on every PR

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#976] [Q&A] [answered]
Why do we fuzz on every PR, isn't doing so inefficient? Would it not be better to just fuzz once a day to see if anything shows up?

(Caveat: I know very little about fuzzing.)

**Comment by apoelstra** [comment, marked as answer]
There are generally two ways that bugs show up during fuzz-tests:
* almost immediately
* Poisson-distributed as you run the fuzztests basically forever

The latter category of bugs would be found more efficiently if we did less-frequent long-running fuzz tests, yes, but (a) I don't know how to do this on Github's servers, while doing short tests on every PR is easy ;), and (b) I don't think that running every day vs every PR is that big of a difference.

Running on every PR is a big benefit for the first category of bugs, but the majority of PRs aren't susceptible to them.

Ideally we'd have somebody running the fuzztests 24/7 for some weeks or months. I've done a couple weeks a few times in the past, but then I always find some other reason that I need the CPU time.

**Reply by tcharding** [reply]
No reason I couldn't spin up a cheap cloud node and run them. I've been wanting to put more time into focusing on testing, probs after the current champion tasks I have in the pipeline. If there are any other areas you think of, related to testing, that need attention, feel free to suggest them. I'll start a tracking issue when I'm getting closer to starting on the work.

**Reply by apoelstra** [reply]
I'm doubtful it'd be valuable on a cheap cloud node. You need to pin the CPU continuously.

**Reply by tcharding** [reply]
oh, scratch that then. Cheers.

**Reply by Kixunil** [reply]
Could we do some crowdfunding to setup a proper rig for this? It shouldn't be a crazy amount of money so I think it's achievable.

**Reply by TheBlueMatt** [reply]
I'll ask the Chaincode folks, they have some fuzzing CI set up running rust-lightning and Bitcoin Core fuzzers already, though we may need another server or two to get them more CPUs. There's also the Google free OSS fuzzing stuff, but you have to apply, dunno how easy that is.

**Reply by apoelstra** [reply]
I'll also try to take some time this month to see if I can't set up some fuzzing infra on my system that (say) only runs at night so it doesn't bug me when I'm working.

**Reply by dpc** [reply]
```
> cat ~/bin/sonice 
#!/bin/sh

exec ionice -c 3 nice -n 20 "$@"
```

Very useful to run heavy stuff "in the background".

**Reply by Kixunil** [reply]
@dpc yes, nice is nice :). It just still runs 100% CPU implying fans at high speed and related noise which can be the issue for some people.

**Comment by tcharding** [comment]
We no longer fuzz on every PR, we have a github action `cron-daily-fuzz.yml`.

**Reply by apoelstra** [reply]
Furthermore, the bitcoinfuzz project has started fuzzing this project continuously it seems. (Or at least, they have for rust-miniscript.)
